### PR TITLE
[Feature] Add runtime_guard control plane (detect, report, dump_kv)

### PR DESCRIPTION
**Pull request by djh95**
Hot-reloadable runtime_config plus wave-synced detectors, report, and KV dump on Ascend v1/v2 runners, with docs and unit tests.
---

### 1. 摘要 / Summary

**中文**:为 vllm-ascend 引入 runtime_guard 控制面:一组低开销在线异常检测器 + 事件报告 + 按请求 KV dump + 热更新配置。目标是在"静默输出异常"与"进程崩溃"之间补上运行时可观测与取证能力——NaN/Inf logits、退化复读、禁词子串、投机解码接受率异常均可在服务不中断的情况下被捕获、报告并转储现场。**默认关闭、零侵入**(C1 实测开启代码不加载检测时开销为 0)。

**English**: Adds the runtime_guard control plane to vllm-ascend: a set of low-overhead online anomaly detectors, incident reporting, per-request KV dumps, and hot-reloadable runtime configuration. It closes the observability gap between "silently wrong output" and "process crash" — NaN/Inf logits, degenerate repetition, forbidden substrings, and spec-decode acceptance anomalies are captured, reported, and dumped for post-mortem without interrupting serving. **Off by default, zero intrusion** (C1 verified: no measurable overhead when detectors are disabled).

### 2. 改动内容 / What's in this PR

68 files, +15,846 / −180;61 个新文件;**纯 Python,csrc 零改动**(无需重编 .so)。rebase 后内容一致,diff 数字略有平移。
68 files, +15,846 / −180; 61 new files; **pure Python, zero csrc changes** (no .so rebuild needed). Identical content on the rebased tip; diffstat shifts slightly.

| 模块 / Module | 内容 / Content |
|---|---|
| `vllm_ascend/runtime_guard/`(31 文件 / files) | 4 个在线检测器 / detectors: `logits_finite`(NaN/Inf)、`token_repeat`(退化复读)、`output_substring`(禁词子串)、`spec_acceptance`(投机接受率);processor/manager、rank_gate(TP0 检测收敛 + 全 rank dump)、detector placement(ExecScope/LPT 分摊)、incident report(per-rank 路由 + quota/cooldown)、dump_kv action(逐层异步 D2H,payload 含 tp_rank/num_kv_heads)、io_snapshot、wave_tracker、`RG_INJECT` 故障注入(**双门禁:源码级总开关默认 False + 环境变量,出厂构建完全忽略 / double-gated: source-level master switch shipped False + env var; shipped builds ignore it entirely**) |
| `vllm_ascend/runtime_config/`(9 文件 / files) | JSONC 运行时配置 + 热更新(可配 interval,默认 0=关)、原子校验(未知键拒绝并保留旧配置)、TP/PP/DP 安全广播、input_filters、ascend_log 级联 / hot-reload config with atomic validation (unknown-key rejection keeps last-good), TP/PP/DP-safe broadcast, input filters, ascend_log cascade |
| worker 接线 / wiring | `model_runner_v1.py`(+441)、`worker/v2/model_runner.py`(+119)、`worker.py`:v1/v2 双 runner、async-scheduling 安全(pre/post-sample 两阶段采样全 rank 生效) / both runners, async-scheduling-safe (two-phase sampling materializes on all ranks) |
| `tests/ut/` + `runtime_guard/test/` | 153 个 UT,行覆盖 75%(含 5 个注入双门禁 shipped-safety 测试)/ 153 UTs, 75% line coverage (incl. 5 double-gate shipped-safety tests) |

### 3. 设计要点 / Design highlights

- **默认零开销 / Zero overhead when off**:检测器/热更新全部关断时,热路径仅有一次 inactive 早退判断;C1(T1/T0≥0.999)双 runner 实测通过。
  When all detectors and hot-reload are off, the hot path takes a single inactive early-exit check; gate C1 (T1/T0 ≥ 0.999) passed on both runners.
- **rank 收敛执行 / Rank-converged execution**:检测收敛到 last-PP TP0(消除每步 D2H 同步性能损失),dump 各 rank 独立执行;报告按 `rank_dp*_tp*_pp*` 路由,无跨 rank 通信。
  Detection converges to last-PP TP0 (eliminating per-step D2H sync tax); dumps run per-rank; reports route to `rank_dp*_tp*_pp*` with no cross-rank communication.
- **配置热更新安全 / Hot-reload safety**:通过广播保证各 rank 集合通信参与一致;拒绝不正确的 JSON 文件、未知 key、非法 value,并保留旧配置。
  Broadcast keeps collective participation rank-consistent; malformed JSON, unknown keys, and invalid values are rejected while the last-good config is kept.
- **NPU 数据布局适配 / NPU layout handling**:v2 runner 的 StagedWriteTensor 与 UvaBufferWrapper 在 host 侧无数据,v1/v2 统一走 `.gpu` 按需拷贝兜底(block_ids / io_snapshot 已实测非空)。仅 dump 会走此逻辑。
  On the v2 runner, StagedWriteTensor and UvaBufferWrapper carry no host-side copy, so both runners fall back to copying rows on demand from `.gpu` (block_ids / io_snapshot verified non-empty on live NPU). Only dumps ever take this path.

### 4. 性能 / Performance gates(910B4·A2 本机 + 910C·A3 远端,交叉轮换 / cross-rotation)

**中文**:

| 门槛 | 定义 | 阈值 | 结果 |
|---|---|---|---|
| C1 | guard 代码加载但全关 vs 无 guard 基线,T1/T0 | ≥0.999 | v1=1.00291 / v2=1.00411;A3:DSV2-Lite 1.0067、Qwen3-30B-A3B 1.0058 — 全过 |
| C2 | 3s 热更新轮询开销,T2/T1 | ≥0.999 | 0.99920~1.00644,过;~0.4% 税仅在 reload 开启时存在,默认 reload_interval=0 无此税 |
| C3 | 检测器开销,T3/T2 | ≥0.990 | v1=1.01120 / v2=0.99980(≤1%),过 |
| C4 | 功能隔离,输出 bit-identical | bit 一致 | 过:双 runner 各 1 轮(4 状态 × 3 prompt = 12 组输出),temp=0/seed=42 全部 bit 一致 |
| C5 | dump 直接计时(无对比基线,纯测量;逐层异步 D2H + save,不在热路径) | — | 每 rank:1k tok 0.15-0.22s + 0.11s(34MiB);16k 0.16-0.72s + 0.33-0.77s;128k 0.64-1.62s + 2.7-5.5s(2-3.9GiB);v1/v2 同量级 |
| C6 | 短压后 RSS/HBM 泄漏回落,guard 组 vs 裸 base 组对照差分 | 差值 <+10MB RSS / <+512MB HBM | guard +0.2MB/−0.7MB vs base +0.0MB/+1.3MB → 差值 +0.1MB/−2.0MB,过,零残留 |

短压定义:1k-128k LongBench 语料持续压测(guard 组 15 请求/244k tok,base 组 17 请求/507k tok;prompt 带序号前缀防 KV 复用),停压后 settle 30s,再连续采样 300s(27 点)。

**English**:

| Gate | Definition | Bar | Result |
|---|---|---|---|
| C1 | guard loaded but all off vs no-guard baseline, T1/T0 | ≥0.999 | v1=1.00291 / v2=1.00411; A3: DSV2-Lite 1.0067, Qwen3-30B-A3B 1.0058 — all pass |
| C2 | 3s hot-reload polling overhead, T2/T1 | ≥0.999 | 0.99920-1.00644, pass; ~0.4% tax exists only while reload is on, absent at the default reload_interval=0 |
| C3 | detector overhead, T3/T2 | ≥0.990 | v1=1.01120 / v2=0.99980 (≤1%), pass |
| C4 | functional isolation, bit-identical outputs | bit-identical | pass: one full pass per runner (4 states × 3 prompts = 12 outputs), temp=0/seed=42, all bit-identical |
| C5 | dump direct timing (no baseline, measurement only; per-layer async D2H + save, off the hot path) | — | per rank: 1k tok 0.15-0.22s + 0.11s (34MiB); 16k 0.16-0.72s + 0.33-0.77s; 128k 0.64-1.62s + 2.7-5.5s (2-3.9GiB); v1/v2 in the same band |
| C6 | post-short-stress RSS/HBM leakback, guard arm vs bare-base control, differential | diff <10MB RSS / <512MB HBM | guard +0.2MB/−0.7MB vs base +0.0MB/+1.3MB → differential +0.1MB/−2.0MB, pass, zero residue |

Short stress = sustained load with 1k-128k LongBench prompts (guard arm 15 req / 244k tok, base arm 17 req / 507k tok; sequence-numbered prompts defeat KV reuse), settle 30s after load stops, then 300s of sampling (27 points).

### 5. 测试证据 / Test evidence

- **功能 / Functional**:P0 冒烟 12/12;P1 5/5;异常注入 4/4(nan/inf/禁词/复读环);热更新软失败;report/dump quota+cooldown;混部双实例 H-01..06;PD 分离 1P1D 冒烟(P/D 双侧挂 guard,D 侧注入,report schema 与单机一致)。
  P0 smoke 12/12; P1 5/5; fault injection 4/4; hot-reload soft-fail; report/dump quota+cooldown; mixed-instance H-01..06; PD-disaggregation 1P1D smoke (guard on both P and D, injection on D, report schema identical to single node).
- **模型与拓扑 / Models & topologies**:DeepSeek-V2-Lite(MoE)、Qwen3-Coder-30B-A3B(MoE)、Qwen2.5-0.5B/7B/8B;runner v1+v2;TP / TP+EP / PP / DP。
- **稳定性 / Stability**:soak 有效 **24.06h**(目标 24h 达标):8 并发持续压测,累计 **4625 轮热更 0 失败**,心跳全程健康,无 hang/crash。
  24.06h effective soak (24h goal met): 8-concurrency sustained load, 4,625 hot-reload cycles with 0 failures, healthy heartbeats throughout, no hang/crash.
- **UT**:`runtime_guard/test` + `tests/ut/runtime_config` = **153 passed / 75%**(含注入双门禁 gate 测试;新 tip 上 135/135 全绿);全量 `tests/ut` 回归 2065 passed / 1 failed(上游既有问题,merge-base 同样失败)。
  153 passed / 75% (incl. double-gate tests; 135/135 green on the rebased tip); full tests/ut regression 2065 passed / 1 pre-existing upstream failure (reproduced on merge-base).
- **注入安全 / Injection safety**:RG_INJECT 双门禁 e2e 实证:shipped 构建 + 环境变量 → 0 注入日志/0 report;翻转源码开关后 → 注入生效 + token_repeat report 落盘。
  Double-gate e2e verified: shipped build + env var → zero injection; armed source build → injection fires, report lands.

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

File: vllm_ascend/worker/v2/model_runner.py [hunk 1]
```diff
@@ -245,10 +256,73 @@ def sample_tokens(self, grammar_output):
             )
 
         self._restore_replicated_draft_target_states()
-        output = super().sample_tokens(grammar_output)
+
+        # Peek before super() pops it: these refs stay valid after the pop.
+        state = self.execute_model_state
+        input_batch = getattr(state, "input_batch", None) if state is not None else None
```
Comment (zouzy5137) [inline vllm_ascend/worker/v2/model_runner.py]: Check if this line can be replaced with state.input_batch

File: vllm_ascend/worker/v2/model_runner.py [hunk 2]
```diff
@@ -245,10 +256,73 @@ def sample_tokens(self, grammar_output):
...
+        if runtime_guard is None:
+            return _super_sample_tokens(grammar_output)
+
+        def sample_fn() -> SamplePhaseResult:
+            with (
+                wrap_compute_logits_for_pre_sample(self, input_batch)
+                if need_pre_sample_hook(self.runtime_guard)
+                else nullcontext()
+            ):
+                output = _super_sample_tokens(grammar_output)
```
Comment (zouzy5137) [inline vllm_ascend/worker/v2/model_runner.py]: Can we just use super().sample_tokens?

File: vllm_ascend/runtime_guard/test/_helpers.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+from vllm_ascend.runtime_guard.processor import RuntimeGuardProcessor
+
+
+def bare_processor() -> RuntimeGuardProcessor:
+    """Minimal ``RuntimeGuardProcessor`` shell for soft-fail / hook wiring UTs."""
+    p = object.__new__(RuntimeGuardProcessor)
+    p.detectors = MagicMock()
+    p.detectors.after_sample_hot_path = MagicMock(side_effect=RuntimeError("boom"))
+    p.detectors.check_before_sample = MagicMock(side_effect=RuntimeError("boom"))
```
Comment (weijinqian0) [inline vllm_ascend/runtime_guard/test/_helpers.py:31]: Move this to test.

File: vllm_ascend/worker/v2/model_runner.py [hunk 4]
```diff
@@ -322,18 +396,36 @@ def execute_model(
...
-            intermediate_tensors=intermediate_tensors,
-            dummy_run=dummy_run,
-            skip_attn_for_dummy_run=skip_attn_for_dummy_run,
-            is_profile=is_profile,
-            context_len=context_len,
-            **({} if vllm_version_is("0.29.0") else {"valid_dummy_state_slots": valid_dummy_state_slots}),
-        )
-        self.model_state.kvpp_is_dummy_run = False
-        self.kvpp.complete_forward()
+        # v1 parity: dummy waves never arm — manual_dump is not burned and the
```
Comment (weijinqian0) [inline vllm_ascend/worker/v2/model_runner.py:399]: Please use annotations to implement.

File: vllm_ascend/ascend_config.py [hunk 5]
```diff
@@ -1650,6 +1654,51 @@ def init_ascend_config(vllm_config: VllmConfig) -> AscendConfig:
     # pre-step; the resolved path is passed as the dump_config_path field.
     dump_config_path = AscendConfig._resolve_dump_config_path(additional_config)
 
+    raw_runtime_path = additional_config.get("runtime_config_path") or additional_config.get("runtime-config")
```
Comment (weijinqian0) [inline vllm_ascend/ascend_config.py:1657]: Extract individual functions
